cdap-app-fabric/src/main/java/co/cask/cdap/internal/profile/ProfileService.java

```diff
@@ -155,23 +169,73 @@ public void createIfNotExists(ProfileId profileId, Profile profile) {
   }
 
   /**
-   * Deletes the profile from the profile store
+   * Deletes the profile from the profile store. Profile deletion must satisfy the following:
+   * 1. Profile must exist and must be DISABLED
+   * 2. There must entities assigned to this profile, now the assignment comes from Preference Store and Schedule store
+   * 3. There must be no active program runs using this profile
    *
    * @param profileId the id of the profile to delete
    * @throws NotFoundException if the profile is not found
    * @throws ProfileConflictException if the profile is enabled
    */
   public void deleteProfile(ProfileId profileId) throws NotFoundException, ProfileConflictException {
     Transactionals.execute(transactional, context -> {
-      getProfileDataset(context).deleteProfile(profileId);
+      ProfileDataset profileDataset = getProfileDataset(context);
+      Profile profile = profileDataset.getProfile(profileId);
+      if (profile == null) {
+        throw new NotFoundException(profileId);
+      }
+
+      // The profile status must be DISABLED
+      if (profile.getStatus() == ProfileStatus.ENABLED) {
+        throw new ProfileConflictException(
+          String.format("Profile %s in namespace %s is currently enabled. A profile can " +
+                          "only be deleted if it is disabled", profileId.getProfile(), profileId.getNamespace()),
+          profileId);
+      }
+
+      // There must be no assigments to this profile
+      Set<EntityId> associations = profileDataset.getProfileAssignment(profileId);
+      if (!associations.isEmpty()) {
+        throw new ProfileConflictException(
+          String.format("There are still entities %s assigned to this profile %s.Please remove the assignment from " +
+                          "the preferenes or schedule properties to delete the profile.",
+                        associations, profileId.toString()),
+          profileId);
+      }
+
+      // There must be no running programs using the profile
+      AppMetadataStore appMetadataStore = AppMetadataStore.create(cConf, context, datasetFramework);
+      NamespaceMDS nsStore = NamespaceMDS.getNamespaceMDS(context, datasetFramework);
+      Set<NamespaceId> namespaceIds;
+      if (profileId.getNamespaceId().equals(NamespaceId.SYSTEM)) {
+        namespaceIds = nsStore.list().stream().map(NamespaceMeta::getNamespaceId).collect(Collectors.toSet());
```

Comment:
instead of getting all namespaces here, it is better to add another getActiveRuns method to AppMetadataStore that doesn't take namespaces. Then you can call that method here and call the namespaced one if this is a user profile.

The run record rowkeys have the state before the namespace, so it can be done in just a single scan.